loom/engine/bindings/loom/lmGameController.cpp

```diff
@@ -0,0 +1,350 @@
+{
+    for (int i = 0; i < MAX_CONTROLLERS; ++i)
+    {
+        LoomGameController::controllers[i].close();
+    }
+    numControllers = 0;
+}
+
+/** Opens a game controller from the game controller pool.
+*  This is usually used with the SDL_CONTROLLERDEVICEADDED event.
+*  Device index can be read from an SDL event using `event.cdevice.which` */
+int LoomGameController::addDevice(int deviceID)
+{
+    // Iterate through the game controller pool and open a controller if a slot is free
+    for (int i = 0; i < MAX_CONTROLLERS; ++i)
+    {
+        LoomGameController& gc = controllers[i];
+        if (!gc.is_connected)
+        {
+            gc.open(deviceID);
+            ++numControllers;
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+/** Closes a game controller in the game controller pool.
+*  This is usually used with the SDL_CONTROLLERDEVICEREMOVED event.
+*  Device index can be read from an SDL event using `event.cdevice.which` */
+int LoomGameController::removeDevice(int deviceID)
+{
+    // We need to get device's index in the game controller pool first in order to close it.
+    int controllerIndex = getControllerIndex(deviceID);
+    if (controllerIndex < 0) return -1;
+    LoomGameController& gc = controllers[controllerIndex];
+    gc.close();
+    --numControllers;
+    return controllerIndex;
+}
+
+/** Returns the game controller's index in the game controller pool using the device's id. */
+int LoomGameController::getControllerIndex(SDL_JoystickID instance)
+{
+    for (int i = 0; i < MAX_CONTROLLERS; i++)
+    {
+        if (LoomGameController::controllers[i].is_connected && LoomGameController::controllers[i].instance_id == instance) {
+            return i;
+        }
+    }
+    return -1;
+}
+
+/** Returns the pointer to a game controller on a specific index in the 'controllers' pool.
+ * If index is negative it returns the first connected game controller in pool. */
+LoomGameController *LoomGameController::getGameController(int index = -1)
+{
+    if (index < 0)
+    {
+        for (int i = 0; i < MAX_CONTROLLERS; i++)
+            if (LoomGameController::controllers[i].is_connected)
+                return &controllers[i];
+    }
+    else if (index < MAX_CONTROLLERS)
+    {
+        if (LoomGameController::controllers[index].is_connected)
+            return &controllers[index];
+    }
+    return nullptr;
+}
+
+/** Returns true if device has rumble support. */
+bool LoomGameController::isHaptic()
+{
+    return this->is_haptic;
+}
+
+/** Stops the device's rumble effect. */
+void LoomGameController::stopRumble()
+{
+    // Do nothing if it is not a valid controller
+    if (!this->is_connected) return;
+
+    // If device is haptic, stop rumble
+    if (this->is_haptic)
+        SDL_HapticRumbleStop(this->getHaptic());
+}
+
+/** Starts the device's rumble effect.
+*  Intensity sets the  is a value between 0 and 1
+*  The duration of rumble is set in milliseconds. Using this */
```

Comment:
Ah, I think I wanted to write some extra stuff and changed my mind later. Forgot to delete it.